persistence-api/src/main/java/io/stargate/db/cdc/CDCServiceImpl.java

```diff
@@ -0,0 +1,143 @@
+
+  private static final CompletableFuture<Void> completedFuture =
+      CompletableFuture.completedFuture(null);
+  private static final CompletableFuture<Void> unhealthyFuture = new CompletableFuture<>();
+  private static final Logger logger = LoggerFactory.getLogger(CDCService.class);
+
+  private final CDCProducer producer;
+  private final CDCHealthChecker healthChecker;
+  private final CDCConfig config;
+  private final ScheduledExecutorService timeoutScheduler =
+      Executors.newScheduledThreadPool(
+          1, new ThreadFactoryBuilder().setDaemon(true).setNameFormat("failAfter-%d").build());
+
+  static {
+    unhealthyFuture.completeExceptionally(
+        new CDCWriteException("CDC producer marked as unhealthy"));
+  }
+
+  public CDCServiceImpl(CDCProducer producer, Metrics metrics) {
+    // TODO: Wrap configuration and use default health checker
+    this(producer, null, null, metrics.getRegistry("cdc"));
+  }
+
+  @VisibleForTesting
+  public CDCServiceImpl(
+      CDCProducer producer,
+      CDCHealthChecker healthChecker,
+      CDCConfig config,
+      MetricRegistry registry) {
+    this.producer = producer;
+    this.healthChecker = healthChecker;
+    this.config = config;
+
+    CDCMetrics.instance.init(registry);
+  }
+
+  @Override
+  public CompletableFuture<Void> publish(MutationEvent mutation) throws CDCWriteException {
+    if (!config.isTrackedByCDC(mutation)) {
+      return completedFuture;
+    }
+
+    if (!healthChecker.isHealthy()) {
+      return unhealthyFuture;
+    }
+
+    return sendToProducer(mutation);
+  }
+
+  private CompletableFuture<Void> sendToProducer(MutationEvent mutation) {
+    final long start = System.nanoTime();
+    CDCMetrics.instance.incrementInFlight();
+    return orTimeout(producer.publish(mutation), config.getProducerTimeoutMs())
+        .whenComplete(
+            (r, e) -> {
+              CDCMetrics.instance.decrementInFlight();
+              CDCMetrics.instance.updateLatency(System.nanoTime() - start);
+
+              if (e != null) {
+                healthChecker.reportSendError(mutation, e);
+                CDCMetrics.instance.markProducerFailure();
+
+                if (e instanceof TimeoutException) {
+                  CDCMetrics.instance.markProducerTimedOut();
+                }
+              } else {
+                healthChecker.reportSendSuccess();
+              }
+            });
+  }
+
+  /** Equivalent of Java 9+ orTimeout() */
+  private <T> CompletableFuture<T> orTimeout(CompletableFuture<T> f, long timeoutMs) {
+    if (timeoutMs <= 0) {
+      return f;
+    }
+
+    try {
+      final ScheduledFuture<Boolean> scheduledTimeout =
+          timeoutScheduler.schedule(
+              () ->
+                  f.completeExceptionally(
+                      new TimeoutException(String.format("Timed out after %sms", timeoutMs))),
+              timeoutMs,
+              MILLISECONDS);
+
+      f.thenAccept(r -> scheduledTimeout.cancel(false));
```

Comment:
shouldn't we return the result of `thenAccept`?

Reply:
I think the caller of this method shouldn't care about the result of `scheduledTimeout.cancel()`. But if cancelling in the background feels strange, I can change it.

Reply:
good point :+1: